Automattic/wp-calypso
# Support more complex pricing displays in one-click checkout modal

#### Proposed Changes

Follow up to pdgrnI-1dv-p2#comment-2300

There has been some discussions around the one-click checkout in the Professional Email post-checkout upsell and whether the line item price in there sometimes displays incorrectly. The assumption was that this problem would impact domains that aren't eligible for the regular introductory offer. This means that there's no immediate concern for regular users, given that this upsell is only triggered if there is an eligible domain at hand after the user has completed a purchase.

However, something else that I came across when testing the one-click checkout is that the product line item renders quite differently from the regular checkout page. So I took a stab at making the following changes:

- If the total cost of the cart is 0, the `Buy` button should read `Complete Checkout` instead of `Pay $0`.
- If the product has an active introductory offer, render the same information as we would in the regular checkout.
- If the product has a sublabel, render the same information as we would in the regular checkout.
- If there is a credits line item in the cart, render that in the price summary.

| Old | New |
| - | - |
|  ![01](https://user-images.githubusercontent.com/1101677/184320913-d80f3c2e-e97a-4d7e-8aa5-902a6ed490fb.png) |  ![02](https://user-images.githubusercontent.com/1101677/184320940-c28ddbbf-48c7-4186-9ac7-5b8696bc9a47.png) |
| ![03](https://user-images.githubusercontent.com/1101677/184321054-f9ae834a-ca59-49c0-b1bd-55237c5dd21a.png) | ![04](https://user-images.githubusercontent.com/1101677/184321087-203ca997-f4a9-4b19-be3e-2c027c39dd5b.png) |

#### Testing Instructions



First and foremost, this should be tested with a domain that has a valid introductory offer:

1. Have a fresh custom domain without an email solution
2. Have a valid payment card stored on your current account (you can check this by navigating to `/me/purchases/payment-methods`)
3. Navigate to `/checkout/offer-professional-email/:domain/123456789/:site_slug` (`123456789` would normally be the receipt ID, but for our purposes, it can be any number)
4. Enter an email address and a password, click `Add Professional Email`
5. Ensure that a one-click checkout modal appears (see above screenshots for reference)
6. Ensure that the sublabel is displayed (`Mailboxes: billed annually`)
7. Ensure that the introductory offer text is displayed (`First 3 months free`)
8. Ensure that the checkout button reads `Complete Checkout`

Since these changes also increase the fault tolerance in the one-click checkout modal, it doesn't hurt to also test with a domain that doesn't have a valid introductory offer. Previously having purchased a Titan subscription and then having removed it is one way to make the introductory offer invalid. To test a domain like that, follow these steps:

1. Have a custom domain that for some reason (see above) doesn't have an introductory offer for Titan
2. Have a valid payment card stored on your current account
3. Navigate to `/checkout/offer-professional-email/:domain/123456789/:site_slug`
4. Enter an email address and a password, click `Add Professional Email`
5. Ensure that a one-click checkout modal appears (see above screenshots for reference)
6. Ensure that the sublabel is displayed (`Mailboxes: billed annually`)
7. Ensure that the introductory offer text is displayed (`Order not eligible for introductory discount`)

#### Pre-merge Checklist






Related to #
